File: packages/typescript/src/toolkit/response.ts
import { BriaException } from "./errors.js";
import { Status, type BriaError, type BriaResult } from "./models.js";

/** A parsed Bria API response. Mirrors the Python `BriaResponse`. */
export class BriaResponse {
  readonly requestId: string;
  readonly status: Status;
  readonly error: BriaError | null;
  readonly result: BriaResult | null;
  readonly statusUrl: string | null;
  readonly headers: Record<string, string>;

  constructor(fields: {
    requestId?: string;
    status: Status;
    error?: BriaError | null;
    result?: BriaResult | null;
    statusUrl?: string | null;
    headers?: Record<string, string>;
  }) {
    this.requestId = fields.requestId ?? "unknown";
    this.status = fields.status;
    this.error = fields.error ?? null;
    this.result = fields.result ?? null;
    this.statusUrl = fields.statusUrl ?? null;
    this.headers = fields.headers ?? {};
  }

  /**
   * Build a response from a raw JSON body. Status is taken from an explicit `status` field if
   * present; otherwise derived: error -> ERROR, else result -> COMPLETED, else status_url ->
   * IN_PROGRESS, else UNKNOWN. (Matches Python `_prepare_model`.)
   */
  static fromBody(
    body: Record<string, unknown>,
    headers: Record<string, string> = {},
  ): BriaResponse {
    const error = (body.error as BriaError | undefined) ?? null;
    const result = (body.result as BriaResult | undefined) ?? null;
    const statusUrl = (body.status_url as string | undefined) ?? null;

    let derived = Status.UNKNOWN;
    if (error !== null) derived = Status.FAILED;
    else if (result !== null) derived = Status.COMPLETED;
    else if (statusUrl !== null) derived = Status.RUNNING;

    const status = (body.status as Status | undefined) ?? derived;

    return new BriaResponse({
      requestId: (body.request_id as string | undefined) ?? "unknown",
      status,
      error,
      result,
      statusUrl,
      headers,
    });
  }

  static fromError(error: BriaError, headers: Record<string, string> = {}): BriaResponse {
    return new BriaResponse({ status: Status.FAILED, error, headers });
  }

  /**
   * Build a response from a completed HTTP response. If the body is not a JSON object, or the
   * HTTP status is >= 400 with no `error` in the body, a `BriaError` is synthesized from the
   * status/reason/text. (Matches Python `from_http_response`.)
   */
  static fromHttpResponse(input: {
    statusCode: number;
    reasonPhrase: string;
    body: unknown;
    bodyText: string;
    headers?: Record<string, string>;
  }): BriaResponse {
    const headers = input.headers ?? {};
    let parsed: BriaResponse | null = null;
    if (input.body !== null && typeof input.body === "object" && !Array.isArray(input.body)) {
      parsed = BriaResponse.fromBody(input.body as Record<string, unknown>, headers);
    }
    if (parsed === null || (input.statusCode >= 400 && parsed.error === null)) {
      return BriaResponse.fromError(
        {
          code: input.statusCode,
          message: input.reasonPhrase || `HTTP ${input.statusCode}`,
          details: input.bodyText,
        },
        headers,
      );
    }
    return parsed;
  }

  /** True while the job is still running. */
  get inProgress(): boolean {
    return this.status === Status.RUNNING;
  }

  /** Throw a `BriaException` if this response carries an error; otherwise no-op. */
  raiseForStatus(): void {
    if (this.error !== null) {
      throw BriaException.fromError(this.error);
    }
  }
}

File: packages/python/src/bria_client/toolkit/response.py
import logging
from typing import Any, NoReturn

from httpx import Response
from pydantic import BaseModel, ConfigDict, Field, ValidationError, model_serializer, model_validator
from pydantic_core.core_schema import SerializationInfo, SerializerFunctionWrapHandler

from bria_client.toolkit.models import BriaError, BriaResult, Status

logger = logging.getLogger(__name__)


class BriaResponse(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    request_id: str = "unknown"
    status: Status
    error: BriaError | None = Field(default=None)
    result: BriaResult | None = Field(default=None)
    status_url: str | None = Field(default=None)
    headers: dict[str, str] = Field(default_factory=dict, exclude=True)

    @model_validator(mode="before")
    @classmethod
    def _prepare_model(cls, data: Any) -> Any:
        status = Status.UNKNOWN
        if data.get("error") is not None:
            status = Status.FAILED
        elif data.get("result") is not None:
            status = Status.COMPLETED
        elif data.get("status_url") is not None:
            status = Status.RUNNING
        data["status"] = data.get("status", status)
        return data

    # noinspection PyUnusedLocal
    @model_serializer(mode="wrap")
    def _force_exclude_none(self, handler: SerializerFunctionWrapHandler, info: SerializationInfo) -> dict[str, Any]:
        result = handler(self)
        # force exclude_none=True over BaseModel
        result = {k: v for k, v in result.items() if v is not None}
        return result

    def __str__(self) -> str:
        # The reason for is to exclude none from str
        return f"<{self.__class__.__name__} {self.model_dump()}>"

    def __repr__(self) -> str:
        # The reason for is to exclude none from repr
        return f"<{self.__class__.__name__} {self.model_dump()}>"

    @classmethod
    def from_error(cls, error: BriaError, headers: dict[str, str] | None = None) -> "BriaResponse":
        return cls(status=Status.FAILED, error=error, headers=headers or {})

    @classmethod
    def from_http_response(cls, response: Response) -> "BriaResponse":
        headers = dict(response.headers)
        try:
            if isinstance(response.json(), dict):
                parsed = cls(**response.json(), headers=headers)
            else:
                raise ValueError("Response is not a JSON object")
        except (ValueError, ValidationError):
            logger.debug("Failed to parse response as BriaResponse")
            parsed = None
        if parsed is None or (response.status_code >= 400 and parsed.error is None):
            return cls.from_error(cls._error_from_response(response), headers=headers)
        return parsed

    @staticmethod
    def _error_from_response(response: Response) -> BriaError:
        return BriaError(
            code=response.status_code,
            message=response.reason_phrase or f"HTTP {response.status_code}",
            details=response.text,
        )

    def raise_for_status(self) -> NoReturn | None:
        if self.error is not None:
            raise self.error.throw()

    @property
    def in_progress(self) -> bool:
        return self.status is Status.RUNNING.value
